### PR TITLE
First wave of creating a PyTorch arm from the TensorFlow code

### PR DESCRIPTION
**Pull request by EtienneChollet**
This pull request introduces the first wave of foundational components for integrating PyTorch support into the neurite library. The primary changes include:

## `BasicUNet`
Introducing `neurite`'s very own UNet architecture with many degrees of customization! `BasicUNet` is a flexible neural network architecture tailored for tasks such as image segmentation--the structure leverages reusable layers and modules from other components of the `neurite` library, enabling streamlined and modular construction of complex models.

### Key Features
* **Supports** 1D, 2D, and 3D data, making it versatile for various applications in medical imaging, computer vision, and more
* **Configurability**: Offers extensive configuration options, allowing users to customize the number of features, kinds of activations, and kinds of normalizations on a per-layer basis (activations and normalizations can be defined globally for the whole network). Furthermore, the `order` parameter of the `BasicUNet` allows users to define the operations in each `ConvBlock` of the unet in any arbitrary order.

## `ConvBlock`
`ConvBlock` is a fundamental network building block of `neurite`, comprising a convolutional layer, and optionally, an activation function and normalization. There will be more features to come and things to include within `ConvBlock`! Stay tuned! With this object, it is possible to define the operations in any arbitrary order by supplying strings of characters to the `order` argument of the constructor.

```mermaid
flowchart LR

subgraph "order='cna'"
    direction LR
    A1((Input)) --> C1[Convolution]
    C1 --> N1[Normalization]
    N1 --> X1[Activation]
    X1 --> B1((Output))
end

subgraph "order='cnacna'"
    direction LR
    A2((Input)) --> C2_1[Convolution]
    C2_1 --> N2_1[Normalization]
    N2_1 --> X2_1[Activation]
    X2_1 --> C2_2[Convolution]
    C2_2 --> N2_2[Normalization]
    N2_2 --> X2_2[Activation]
    X2_2 --> B2((Output))
end

subgraph "order='ccccc'"
    direction LR
    A3((Input)) --> C3_1[Convolution]
    C3_1 --> X3_1[Convolution]
    X3_1 --> C3_2[Convolution]
    C3_2 --> C3_3[Convolution]
    C3_3 --> C3_4[Convolution]
    C3_4 --> B3((Output))
end
```

## `neurite/torch/layers.py`

Implements PyTorch equivalents of existing TensorFlow layers, such as `Resize`, `SoftQuantize`, and `MSE` .

- Upgrades many transforms such as `Resize` , `SoftQuantize`, and `GaussianBlur` (and their functional counterparts) by accepting samplers (and of course, fixed values) as arguments. For instance, the `sigma` parameter of `GaussianBlur` can now be a uniformly distributed random value by specifying e.g. `GaussianBlur(sigma=Uniform(4, 5))`.
   - With all of the other cool samplers that we now have in neurite, lets see what else we can define sigma to be!:
      - `GaussianBlur(sigma=LogNormal(mean=1, variance=0.3))`
      - `GaussianBlur(sigma=Normal(2, 1))`
      - `GaussianBlur(sigma=RandInt(1, 5))` 

## `neurite/torch/random.py`

Introduces a lot of statistical samplers for various stochastic processes.

- Adds support for defining random variables with distributions such as Uniform, Normal, and other customizable samplers, enabling dynamic and flexible parameterization across transforms.
- Facilitates the incorporation of randomness into PyTorch layers and transforms, allowing parameters like `sigma` or `scale` to vary dynamically during training or inference.
- Ensures that random samplers are serializable and compatible with Neurite's broader infrastructure, enhancing reproducibility and ease of saving/loading models.

### Example

```python
# Import the Uniform distribution. I've made many more: Poisson, Bernoulli, Normal, etc...
>>> from neurite.torch.random import Uniform
# Initialize the sampler on an inclusive range
>>> sampler = Uniform(13, 14)
# Let's see how it looks serialized
>>> sampler.serialize()
{
   'qualname': 'Uniform',
   'parent': 'Sampler',
   'module': 'neurite.torch.random',
   'theta': {
      'min_val': 13,
      'max_val': 14
      }
   }
```

## **`neurite/torch/utils.py`**

Provides utility functions to support PyTorch integration?

- Introduces helper functions for common operations

### CODE REVIEW

**Comment by EtienneChollet** [comment]
Make code (especially in `utils.py` more breathable

**Comment by EtienneChollet** [comment]
Make code (especially in utils.py more breathable

**Comment by EtienneChollet** [comment]
> Make code (especially in utils.py more breathable

Resolved in 52ceb7a1d8001f30fdf6e70f3df75d6906d2e813, ede1a01a3a18dcd6b783f4fa5eb597db9661015d
